File: platform-services/gtm-swarm/src/task_queue.py
#!/usr/bin/env python3
"""
Unison Orchestration — Phase 2 Pillar 1 Commit 2 Task Queue
Async background execution tracking in shared .agent_state SQLite cluster.
"""

from __future__ import annotations

import sqlite3
import time
import uuid
from pathlib import Path
from typing import Any

from sqlite_elite import CONNECT_TIMEOUT_SEC, apply_elite_pragmas

def _default_db() -> Path:
    from state_paths import agent_memory_db, ensure_state_dirs

    ensure_state_dirs()
    return agent_memory_db()


_DEFAULT_DB = _default_db()

_VALID_STATUSES = frozenset(
    {"pending", "running", "completed", "failed", "cancelled"}
)


class TaskQueueStore:
    """Durable async task queue for coordinator daemon ticks."""

    def __init__(self, db_path: str | Path | None = None) -> None:
        self.db_path = Path(db_path or _DEFAULT_DB)
        self.db_path.parent.mkdir(parents=True, exist_ok=True)
        self._init_schema()

    def _connect(self) -> sqlite3.Connection:
        conn = sqlite3.connect(str(self.db_path), timeout=CONNECT_TIMEOUT_SEC)
        conn.row_factory = sqlite3.Row
        apply_elite_pragmas(conn)
        return conn

    def _init_schema(self) -> None:
        with self._connect() as conn:
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS task_queue (
                    task_id TEXT PRIMARY KEY,
                    agent_id TEXT NOT NULL,
                    session_id TEXT NOT NULL,
                    collection TEXT NOT NULL,
                    query TEXT NOT NULL,
                    status TEXT NOT NULL DEFAULT 'pending',
                    created_at REAL NOT NULL,
                    completed_at REAL,
                    result_digest TEXT,
                    workflow_dsl TEXT
                )
                """
            )
            try:
                conn.execute(
                    "ALTER TABLE task_queue ADD COLUMN workflow_dsl TEXT"
                )
            except sqlite3.OperationalError:
                pass
            conn.execute(
                """
                CREATE INDEX IF NOT EXISTS idx_task_queue_status_created
                ON task_queue (status, created_at ASC)
                """
            )
            conn.commit()

    def enqueue_task(
        self,
        agent_id: str,
        session_id: str,
        collection: str,
        query: str,
        workflow_dsl: str | None = None,
    ) -> str:
        aid = agent_id.strip()
        sid = session_id.strip()
        col = collection.strip()
        q = query.strip()
        if not aid or not sid or not col or not q:
            raise ValueError("agent_id, session_id, collection, and query are required")

        task_id = str(uuid.uuid4())
        now = time.time()
        with self._connect() as conn:
            conn.execute(
                """
                INSERT INTO task_queue
                (task_id, agent_id, session_id, collection, query, status,
                 created_at, completed_at, result_digest, workflow_dsl)
                VALUES (?, ?, ?, ?, ?, 'pending', ?, NULL, NULL, ?)
                """,
                (task_id, aid, sid, col, q, now, workflow_dsl),
            )
            conn.commit()
        return task_id

    def fetch_next_pending_task(self) -> dict[str, Any] | None:
        """Atomically claim the oldest pending task (pending → running)."""
        with self._connect() as conn:
            conn.execute("BEGIN IMMEDIATE")
            row = conn.execute(
                """
                SELECT task_id, agent_id, session_id, collection, query, status,
                       created_at, completed_at, result_digest, workflow_dsl
                FROM task_queue
                WHERE status = 'pending'
                ORDER BY created_at ASC
                LIMIT 1
                """
            ).fetchone()
            if row is None:
                conn.execute("ROLLBACK")
                return None

            now = time.time()
            conn.execute(
                """
                UPDATE task_queue
                SET status = 'running', completed_at = NULL
                WHERE task_id = ?
                """,
                (row["task_id"],),
            )
            conn.commit()
            return {
                "task_id": row["task_id"],
                "agent_id": row["agent_id"],
                "session_id": row["session_id"],
                "collection": row["collection"],
                "query": row["query"],
                "status": "running",
                "created_at": row["created_at"],
                "completed_at": None,
                "result_digest": None,
                "workflow_dsl": row["workflow_dsl"],
                "claimed_at": now,
            }

    def get_task(self, task_id: str) -> dict[str, Any] | None:
        tid = task_id.strip()
        if not tid:
            return None
        with self._connect() as conn:
            row = conn.execute(
                """
                SELECT task_id, agent_id, session_id, collection, query, status,
                       created_at, completed_at, result_digest, workflow_dsl
                FROM task_queue
                WHERE task_id = ?
                """,
                (tid,),
            ).fetchone()
        if row is None:
            return None
        return dict(row)

    def update_task_status(
        self,
        task_id: str,
        status: str,
        result_digest: str | None = None,
    ) -> dict[str, Any] | None:
        tid = task_id.strip()
        st = status.strip().lower()
        if not tid:
            raise ValueError("task_id is required")
        if st not in _VALID_STATUSES:
            raise ValueError(f"invalid status: {status}")

        completed_at: float | None = None
        if st in {"completed", "failed", "cancelled"}:
            completed_at = time.time()

        with self._connect() as conn:
            cur = conn.execute(
                """
                UPDATE task_queue
                SET status = ?,
                    result_digest = COALESCE(?, result_digest),
                    completed_at = COALESCE(?, completed_at)
                WHERE task_id = ?
                """,
                (st, result_digest, completed_at, tid),
            )
            if cur.rowcount == 0:
                conn.commit()
                return None
            conn.commit()
        return self.get_task(tid)


def enqueue_task(
    agent_id: str,
    session_id: str,
    collection: str,
    query: str,
    *,
    db_path: str | Path | None = None,
) -> str:
    return TaskQueueStore(db_path).enqueue_task(
        agent_id, session_id, collection, query
    )


def fetch_next_pending_task(
    *,
    db_path: str | Path | None = None,
) -> dict[str, Any] | None:
    return TaskQueueStore(db_path).fetch_next_pending_task()


def update_task_status(
    task_id: str,
    status: str,
    result_digest: str | None = None,
    *,
    db_path: str | Path | None = None,
) -> dict[str, Any] | None:
    return TaskQueueStore(db_path).update_task_status(
        task_id, status, result_digest
    )


def get_task(
    task_id: str,
    *,
    db_path: str | Path | None = None,
) -> dict[str, Any] | None:
    return TaskQueueStore(db_path).get_task(task_id)

File: core-mcp-server/src/task_queue.rs
//! Async task queue — durable background execution tracking on Fly NVMe volumes.

use std::path::{Path, PathBuf};
use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};

use rusqlite::{params, Connection, OptionalExtension};
use thiserror::Error;
use tracing::{info, warn};
use uuid::Uuid;

pub const DEFAULT_TASK_DB_PATH: &str = "/data/task_queue.db";

#[derive(Debug, Error)]
pub enum TaskQueueError {
    #[error("sqlite error: {0}")]
    Sqlite(#[from] rusqlite::Error),
    #[error("mutex poisoned")]
    Poisoned,
    #[error("io error: {0}")]
    Io(#[from] std::io::Error),
    #[error("invalid input: {0}")]
    InvalidInput(String),
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct TaskRecord {
    pub task_id: String,
    pub agent_id: String,
    pub session_id: String,
    pub collection: String,
    pub query: String,
    pub status: String,
    pub created_at: f64,
    pub completed_at: Option<f64>,
    pub result_digest: Option<String>,
}

pub struct TaskQueueStore {
    path: PathBuf,
    conn: Mutex<Connection>,
}

impl TaskQueueStore {
    pub fn open(path: impl AsRef<Path>) -> Result<Self, TaskQueueError> {
        let path = path.as_ref().to_path_buf();
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        let conn = Connection::open(&path)?;
        conn.execute_batch(
            "PRAGMA journal_mode=WAL;
             PRAGMA synchronous=NORMAL;
             PRAGMA busy_timeout=5000;
             CREATE TABLE IF NOT EXISTS task_queue (
               task_id TEXT PRIMARY KEY,
               agent_id TEXT NOT NULL,
               session_id TEXT NOT NULL,
               collection TEXT NOT NULL,
               query TEXT NOT NULL,
               status TEXT NOT NULL DEFAULT 'pending',
               created_at REAL NOT NULL,
               completed_at REAL,
               result_digest TEXT
             );
             CREATE INDEX IF NOT EXISTS idx_task_queue_status_created
               ON task_queue (status, created_at ASC);",
        )?;
        info!("Task queue persistence online: {}", path.display());
        Ok(Self {
            path,
            conn: Mutex::new(conn),
        })
    }

    pub fn resolve_db_path() -> PathBuf {
        std::env::var("TASK_QUEUE_DB_PATH")
            .map(PathBuf::from)
            .unwrap_or_else(|_| PathBuf::from(DEFAULT_TASK_DB_PATH))
    }

    pub fn open_with_fallback() -> Self {
        let primary = Self::resolve_db_path();
        match Self::open(&primary) {
            Ok(store) => return store,
            Err(err) => {
                warn!(
                    "Primary task queue DB unavailable at {} ({err}); using /tmp fallback",
                    primary.display()
                );
            }
        }
        let fallback = PathBuf::from("/tmp/task_queue.db");
        Self::open(&fallback).unwrap_or_else(|e| {
            panic!("Failed to open task queue DB at {}: {e}", fallback.display());
        })
    }

    pub fn db_path(&self) -> &Path {
        &self.path
    }

    pub fn enqueue_task(
        &self,
        agent_id: &str,
        session_id: &str,
        collection: &str,
        query: &str,
    ) -> Result<TaskRecord, TaskQueueError> {
        let agent_id = agent_id.trim();
        let session_id = session_id.trim();
        let collection = collection.trim();
        let query = query.trim();
        if agent_id.is_empty()
            || session_id.is_empty()
            || collection.is_empty()
            || query.is_empty()
        {
            return Err(TaskQueueError::InvalidInput(
                "agent_id, session_id, collection, and query are required".into(),
            ));
        }

        let task_id = Uuid::new_v4().to_string();
        let created_at = unix_now_secs();
        let conn = self.conn.lock().map_err(|_| TaskQueueError::Poisoned)?;
        conn.execute(
            "INSERT INTO task_queue
             (task_id, agent_id, session_id, collection, query, status,
              created_at, completed_at, result_digest)
             VALUES (?1, ?2, ?3, ?4, ?5, 'pending', ?6, NULL, NULL)",
            params![task_id, agent_id, session_id, collection, query, created_at],
        )?;
        Ok(TaskRecord {
            task_id,
            agent_id: agent_id.to_string(),
            session_id: session_id.to_string(),
            collection: collection.to_string(),
            query: query.to_string(),
            status: "pending".to_string(),
            created_at,
            completed_at: None,
            result_digest: None,
        })
    }

    pub fn get_task(&self, task_id: &str) -> Result<Option<TaskRecord>, TaskQueueError> {
        let task_id = task_id.trim();
        if task_id.is_empty() {
            return Ok(None);
        }
        let conn = self.conn.lock().map_err(|_| TaskQueueError::Poisoned)?;
        let row = conn
            .query_row(
                "SELECT task_id, agent_id, session_id, collection, query, status,
                        created_at, completed_at, result_digest
                 FROM task_queue WHERE task_id = ?1",
                params![task_id],
                row_to_task,
            )
            .optional()?;
        Ok(row)
    }

    pub fn fetch_next_pending_task(&self) -> Result<Option<TaskRecord>, TaskQueueError> {
        let conn = self.conn.lock().map_err(|_| TaskQueueError::Poisoned)?;
        conn.execute_batch("BEGIN IMMEDIATE")?;
        let row = conn
            .query_row(
                "SELECT task_id, agent_id, session_id, collection, query, status,
                        created_at, completed_at, result_digest
                 FROM task_queue
                 WHERE status = 'pending'
                 ORDER BY created_at ASC
                 LIMIT 1",
                [],
                row_to_task,
            )
            .optional()?;

        let Some(mut task) = row else {
            conn.execute_batch("ROLLBACK")?;
            return Ok(None);
        };

        conn.execute(
            "UPDATE task_queue SET status = 'running', completed_at = NULL WHERE task_id = ?1",
            params![task.task_id],
        )?;
        conn.execute_batch("COMMIT")?;
        task.status = "running".to_string();
        task.completed_at = None;
        Ok(Some(task))
    }

    pub fn update_task_status(
        &self,
        task_id: &str,
        status: &str,
        result_digest: Option<&str>,
    ) -> Result<Option<TaskRecord>, TaskQueueError> {
        let task_id = task_id.trim();
        let status = status.trim().to_lowercase();
        if task_id.is_empty() {
            return Err(TaskQueueError::InvalidInput("task_id is required".into()));
        }
        if !matches!(
            status.as_str(),
            "pending" | "running" | "completed" | "failed" | "cancelled"
        ) {
            return Err(TaskQueueError::InvalidInput(format!(
                "invalid status: {status}"
            )));
        }

        let completed_at = if matches!(status.as_str(), "completed" | "failed" | "cancelled") {
            Some(unix_now_secs())
        } else {
            None
        };

        let conn = self.conn.lock().map_err(|_| TaskQueueError::Poisoned)?;
        let updated = conn.execute(
            "UPDATE task_queue
             SET status = ?1,
                 result_digest = COALESCE(?2, result_digest),
                 completed_at = COALESCE(?3, completed_at)
             WHERE task_id = ?4",
            params![status, result_digest, completed_at, task_id],
        )?;
        if updated == 0 {
            return Ok(None);
        }
        drop(conn);
        self.get_task(task_id)
    }
}

fn row_to_task(row: &rusqlite::Row<'_>) -> rusqlite::Result<TaskRecord> {
    Ok(TaskRecord {
        task_id: row.get(0)?,
        agent_id: row.get(1)?,
        session_id: row.get(2)?,
        collection: row.get(3)?,
        query: row.get(4)?,
        status: row.get(5)?,
        created_at: row.get(6)?,
        completed_at: row.get(7)?,
        result_digest: row.get(8)?,
    })
}

fn unix_now_secs() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0)
}
